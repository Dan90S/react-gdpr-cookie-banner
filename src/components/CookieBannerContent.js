import React from 'react';
import CookieOption from './CookieOption';
import bannerStyle from './bannerStyle';

export default (props = {}) => {
  const {
    styles = {},
    className = '',
    message = 'No text',
    policyLink = '/#',
    impressumLink = '/#',
    privacyPolicyLinkText = 'Privacy Policy',
    impressumLinkText = 'Data Protection',
    necessaryOptionText = 'Necessary',
    preferencesOptionText = 'Preferences',
    statisticsOptionText = 'Statistics',
    marketingOptionText = 'Marketing',
    showAcceptSelectionButton = false,
    acceptAllButtonText = 'Accept All',
    acceptSelectionButtonText = 'Accept Selection',
    showPreferencesOption = true,
    showStatisticsOption = true,
    showMarketingOption = true,
    preferencesOptionInitiallyChecked = false,
    statisticsOptionInitiallyChecked = false,
    marketingOptionInitiallyChecked = false,
    onTogglePreferencesCookies = Function,
    onToggleStatisticsCookies = Function,
    onToggleMarketingCookies = Function,
    onConfirmSelection = Function,
    onConfirmAll = Function,
  } = props;

  const {
    dialog: dialogStyle,
    container: containerStyle,
    message: messageStyle,
    policy: policyStyle,
    selectPane: selectPaneStyle,
    optionWrapper: optionWrapperStyle,
    optionLabel: optionLabelStyle,
    checkbox: checkboxStyle,
    buttonWrapper: buttonWrapperStyle,
    button: buttonStyle,
  } = { ...bannerStyle, ...styles };

  const cookieOptionStyle = { optionWrapperStyle, optionLabelStyle, checkboxStyle };

  return (
    <div className={`react-cookie-law-dialog ${className}`} style={dialogStyle}>
      <div className="react-cookie-law-container" style={containerStyle}>
        <div className="react-cookie-law-msg" style={messageStyle}>{message}</div>

        <div className="react-cookie-law-select-pane" style={selectPaneStyle}>
          <CookieOption
            id="check-required-cookies"
            text={necessaryOptionText}
            styles={cookieOptionStyle}
            disabled
            checked
          />

          {
            showPreferencesOption && (
              <CookieOption
                id="check-preferences-cookies"
                text={preferencesOptionText}
                styles={cookieOptionStyle}
                onChange={onTogglePreferencesCookies}
                checked={preferencesOptionInitiallyChecked}
              />
            )
          }

          {
            showStatisticsOption && (
              <CookieOption
                id="check-statistics-cookies"
                text={statisticsOptionText}
                styles={cookieOptionStyle}
                onChange={onToggleStatisticsCookies}
                checked={statisticsOptionInitiallyChecked}
              />
            )
          }

          {
            showMarketingOption && (
              <CookieOption
                id="check-marketing-cookies"
                text={marketingOptionText}
                styles={cookieOptionStyle}
                onChange={onToggleMarketingCookies}
                checked={marketingOptionInitiallyChecked}
              />
            )
          }
        </div>

        <a href={policyLink} className="react-cookie-law-policy" style={policyStyle}>{privacyPolicyLinkText}</a>
        <a href={impressumLink} className="react-cookie-law-policy" style={policyStyle}>{impressumLinkText}</a>

        <div className="react-cookie-law-button-wrapper" style={buttonWrapperStyle}>
          {
            showAcceptSelectionButton && (
              <button type="button" className="react-cookie-law-accept-selection-btn" style={buttonStyle} onClick={() => onConfirmSelection()}>
                <span>{acceptSelectionButtonText}</span>
              </button>
            )
          }

          <button type="button" className="react-cookie-law-accept-all-btn" style={buttonStyle} onClick={() => onConfirmAll()}>
            <span>{acceptAllButtonText}</span>
          </button>
        </div>
      </div>
    </div>
  );
};
