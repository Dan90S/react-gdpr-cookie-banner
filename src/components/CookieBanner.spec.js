import React from 'react';
import { mount, shallow } from 'enzyme';
import CookieBanner from './CookieBanner';
import CookieBannerContent from './CookieBannerContent';

describe('CookieBanner component', () => {
  beforeEach((done) => {
    document.cookie.split(';').forEach((cookie) => {
      const eqPos = cookie.indexOf('=');
      const name = eqPos > -1 ? cookie.substr(0, eqPos) : cookie;
      document.cookie = `${name}=;expires=Thu, 01 Jan 1970 00:00:00 GMT`;
    });

    done();
  });

  test('should be rendered', () => {
    const component = shallow(
      <CookieBanner message="Text" />,
    );

    expect(component).toMatchSnapshot();
  });

  test('should be rendered with elements', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    expect(component.find(CookieBannerContent)).toHaveLength(1);
  });


  test('should be rendered with custom class', () => {
    const component = mount(
      <CookieBanner className={'testclass'} message="Custom text" />,
    );

    expect(component.hasClass('testclass')).toBe(true)
  });

  test('shouldn\'t show banner if cookies are already accepted', () => {
    document.cookie = 'rcl_consent_given=true';

    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    expect(component.find(CookieBannerContent)).toHaveLength(0);
  });

  test('should close banner if confirm function is called', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    expect(component.html().length).toBeGreaterThan(0);
    expect(component.contains(CookieBannerContent)).toBeTruthy();
    component.find(CookieBannerContent).prop('onConfirmAll')();
    expect(component.html()).toBeNull();
  });

  test('should close banner if decline button is pressed', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    expect(component.html().length).toBeGreaterThan(0);
    expect(component.contains(CookieBannerContent)).toBeTruthy();
    component.find(CookieBannerContent).prop('onConfirmSelection')();
    expect(component.html()).toBeNull();
  });

  test('should close banner at scrolling if dismissOnScroll is true', () => {
    const component = mount(
      <CookieBanner message="Test" dismissOnScroll />,
    );

    expect(component.html().length).toBeGreaterThan(0);
    window.dispatchEvent(new window.UIEvent('scroll', { detail: 0 }));
    expect(component.html()).toBeNull();
  });

  test('shouldn\'t close banner at scrolling if dismissOnScroll is true', () => {
    const component = mount(
      <CookieBanner message="Test" />,
    );

    expect(component.html().length).toBeGreaterThan(0);
    window.dispatchEvent(new window.UIEvent('scroll', { detail: 0 }));
    expect(component.html().length).toBeGreaterThan(0);
  });

  test('should give consent if confirm function is called', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onConfirmAll')();

    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
  });

  test('should give consent if decline function is called', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onConfirmSelection')();

    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
  });

  test('should give consent at scrolling if dismissOnScroll is true', () => {
    mount(<CookieBanner message="Custom text" dismissOnScroll />);

    window.dispatchEvent(new window.UIEvent('scroll', { detail: 0 }));
    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
  });

  test('should accept all consents at confirm all', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onConfirmAll')();

    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
    expect(document.cookie.indexOf('rcl_preferences_consent=true')).toBeGreaterThanOrEqual(0);
    expect(document.cookie.indexOf('rcl_statistics_consent=true')).toBeGreaterThanOrEqual(0);
    expect(document.cookie.indexOf('rcl_marketing_consent=true')).toBeGreaterThanOrEqual(0);
  });
  
  test('should give consent to marketing only at confirm selected', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onToggleMarketingCookies')();
    component.find(CookieBannerContent).prop('onConfirmSelection')();

    expect(document.cookie).toBe('rcl_consent_given=true; rcl_marketing_consent=true');
  });

  test('should give consent to statistics only at confirm selected', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onToggleStatisticsCookies')();
    component.find(CookieBannerContent).prop('onConfirmSelection')();

    expect(document.cookie).toBe('rcl_consent_given=true; rcl_statistics_consent=true');
  });

  test('should give consent to preferences only at confirm selected', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onTogglePreferencesCookies')();
    component.find(CookieBannerContent).prop('onConfirmSelection')();

    expect(document.cookie).toBe('rcl_consent_given=true; rcl_preferences_consent=true');
  });
});
