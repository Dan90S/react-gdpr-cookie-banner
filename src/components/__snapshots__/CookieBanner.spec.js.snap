// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieBanner component should be rendered 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CookieBanner
    message="Text"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "acceptAllButtonText": undefined,
      "acceptSelectionButtonText": undefined,
      "className": undefined,
      "marketingOptionInitiallyChecked": undefined,
      "marketingOptionText": undefined,
      "message": "Text",
      "necessaryOptionText": undefined,
      "onConfirmAll": [Function],
      "onConfirmSelection": [Function],
      "onToggleMarketingCookies": [Function],
      "onTogglePreferencesCookies": [Function],
      "onToggleStatisticsCookies": [Function],
      "policyLink": undefined,
      "preferencesOptionInitiallyChecked": undefined,
      "preferencesOptionText": undefined,
      "privacyPolicyLinkText": undefined,
      "showAcceptSelectionButton": undefined,
      "showMarketingOption": undefined,
      "showPreferencesOption": undefined,
      "showStatisticsOption": undefined,
      "statisticsOptionInitiallyChecked": undefined,
      "statisticsOptionText": undefined,
      "styles": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "acceptAllButtonText": undefined,
        "acceptSelectionButtonText": undefined,
        "className": undefined,
        "marketingOptionInitiallyChecked": undefined,
        "marketingOptionText": undefined,
        "message": "Text",
        "necessaryOptionText": undefined,
        "onConfirmAll": [Function],
        "onConfirmSelection": [Function],
        "onToggleMarketingCookies": [Function],
        "onTogglePreferencesCookies": [Function],
        "onToggleStatisticsCookies": [Function],
        "policyLink": undefined,
        "preferencesOptionInitiallyChecked": undefined,
        "preferencesOptionText": undefined,
        "privacyPolicyLinkText": undefined,
        "showAcceptSelectionButton": undefined,
        "showMarketingOption": undefined,
        "showPreferencesOption": undefined,
        "showStatisticsOption": undefined,
        "statisticsOptionInitiallyChecked": undefined,
        "statisticsOptionText": undefined,
        "styles": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
